### PR TITLE
[TestLoop] Add a test for syncing, and necessary TestLoop enhancements.

### PR DESCRIPTION
**Pull request by robin-near**
Several improvements:

* Allow a new TestLoop to be constructed with reused Store's and TempDir.
* Allow arbitrary override of node config in the testloop builder (this isn't used in this PR but will be later).
* Networking code no longer uses the very awkward route back global static map. It instead uses a shared state. This also allows us to support routing not just based on AccountId, but also PeerId (which is needed for e.g. BlockHeadersRequest).
* Add ViewClientActor support to the test loop networking layer, which is needed for answering queries e.g. BlockHeadersRequest.
* In the log visualizer UI, support log files that contain multiple testloop runs. When visualizing such a file, the UI will first ask you which run to visualize.

Add a test for the scenario of bootstrapping a new node using header, state, and block sync (though the test doesn't dictate which sync method must be used). This is both a useful test, and also a preparation to add epoch sync later.

### CODE REVIEW

**Comment by frol** [comment]
@race-of-sloths invite

**Comment by race-of-sloths** [comment]
@frol Thank you for calling!
    
@robin-near Thank you for the contribution! Join [Race of Sloths](https://race-of-sloths.com) by simply mentioning me in your comment/PRs description and start collecting Sloth Points through contributions to open source projects.

<details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/robin-near)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Invite contributor** `@race-of-sloths invite` to invite the contributor to participate in a race or include it, if it's already a runner.
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called
</details>

File: integration-tests/src/test_loop/tests/syncing.rs
Comment (Longarithm) [inline integration-tests/src/test_loop/tests/syncing.rs:99]: Consider copying code from `shutdown_and_drain_remaining_events`:
```
for node_data in self.datas {
    self.test_loop.data.get_mut(&node_data.state_sync_dumper_handle).stop();
}

self.test_loop.shutdown_and_drain_remaining_events(timeout);
```

Fake tempdir looks more fragile.

Or make `shutdown_and_drain_remaining_events` return that tempdir, letting caller to decide whether to drop it or not.

File: integration-tests/src/test_loop/builder.rs
Comment (Longarithm) [inline integration-tests/src/test_loop/builder.rs:48]: Please mention this is `(hot_store, split_store)` to make it fully obvious.

File: chain/network/src/test_loop.rs
Comment (Longarithm) [inline chain/network/src/test_loop.rs:117]: Maybe PeerClientSenders, "One" looks a bit redundant otherwise.

File: tools/debug-ui/src/LogFileDrop.tsx
Comment (Longarithm) [inline tools/debug-ui/src/LogFileDrop.tsx:91]: I should probably wait for the demo, but anyway - is it expected to support multiple tests or multiple runs of the same test?
Because in case of the latter, how can I map the test to the log number (I assume that's what `i` is for)?

**Review by Longarithm — APPROVED** [review]
LGTM, leaving some nits.

**Comment by shreyan-gupta** [comment]
I'll review this by Monday. Apologies for delay.

File: tools/debug-ui/src/LogFileDrop.tsx
Comment (robin-near) [reply tools/debug-ui/src/LogFileDrop.tsx:91]: Yes that's what the i is for. The UI will display multile buttons that each says "Log #x"

File: integration-tests/src/test_loop/builder.rs
Comment (robin-near) [reply integration-tests/src/test_loop/builder.rs:63]: Yeah well, I agree on that idea but I'm not sure how to build one right now without introducing more mess. The problem is that the default configs are not very default to begin with. So introducing a builder only gives the illusion that the config is being predictably constructed. I think the better thing to do would be to thin down TestLoopBuilder a bit so that it is more customizable, and then the client config builder would be part of that; for now I don't think ClientConfigBuilder is the best solution (it may end up being unnecessary once TestLoopBuilder is better designed; like enable_state_sync would both enable the config and enable the corresponding functionality in the builder setup code).

File: chain/network/src/test_loop.rs
Comment (shreyan-gupta) [inline chain/network/src/test_loop.rs:124]: Nice abstraction

File: chain/network/src/test_loop.rs
Comment (shreyan-gupta) [inline chain/network/src/test_loop.rs:43]: Curious why did we change this from Arc to Box? Is there some rusty philosophical reason or both work?

**Pull request merged**
